numpy/lib/_function_base_impl.py

```diff
@@ -4535,9 +4535,11 @@ def quantile(a,
         if axis is not None:
             axis = _nx.normalize_axis_tuple(axis, a.ndim, argname="axis")
         weights = _weights_are_valid(weights=weights, a=a, axis=axis)
-        if np.any(weights < 0):
-            raise ValueError("Weights must be non-negative.")
-
+        if np.any(weights <= 0):
+            if np.any(weights < 0):
+                raise ValueError("Weights must be non-negative.")
+            elif np.all(weights == 0):
```

Comment:
can't you just say `else` here?

Reply:
you can move the `== 0` to the sum to simplify this, though.

IMO the checks may be clearer as `not all(weights >= 0)` to reject NaNs right away, but doesn't matter much so long the `== 0` case is handled below.

Also, why not use `isfinite`?